### PR TITLE
routing key determinator

### PR DESCRIPTION
**Pull request by doubret**
Implemented routing key determinator.
Fixes issue #6

### CODE REVIEW

File: lib/metrics/index.ts [hunk 1]
```diff
@@ -10,3 +10,4 @@ export * from "./IHistogramValue";
 export * from "./IMeterValue";
 export * from "./ITimerValue";
 export * from "./MetricMessageBuilder";
+export * from "./RoutingKeyDeterminator";
```
Comment (rstiller) [inline lib/metrics/index.ts]: `AmqpReportingResult` is missing

File: tslint.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
     "interface-name": [false],
     "only-arrow-functions":[false],
     "no-var-requires": false,
-    "max-line-length": [true, 180],
+    "max-line-length": [true, 200],
```
Comment (rstiller) [inline tslint.json]: Don't want to modify / format the code instead?
Not sure what a good, common default value would be ...  I always use 120 ...

File: lib/metrics/AmqpMetricReporter.ts [hunk 3]
```diff
@@ -365,10 +385,10 @@ export class AmqpMetricReporter extends ScheduledMetricReporter<AmqpMetricReport
...
    */
-  protected handleResults(ctx: OverallReportContext, registry: MetricRegistry, date: Date, type: MetricType, results: Array<ReportingResult<any, Message>>): Promise<void> {
+  protected handleResults(ctx: OverallReportContext, registry: MetricRegistry, date: Date, type: MetricType, results: Array<ReportingResult<any, AmqpReportingResult>>): Promise<void> {
     results
-      .filter((result) => result.result)
-      .forEach((result) => result.result.sendTo(this.target));
+      .filter((result) => result.result && result.result.message)
+      .forEach((result) => result.result.message.sendTo(this.target, result.result.routingKey));
 
     return Promise.resolve();
```
Comment (rstiller) [inline lib/metrics/AmqpMetricReporter.ts]: If you use the async keyword before the function name you can skip this line entirely.

File: lib/metrics/AmqpMetricReporter.ts [hunk 4]
```diff
@@ -383,8 +403,12 @@ export class AmqpMetricReporter extends ScheduledMetricReporter<AmqpMetricReport
    * @returns {{}}
    * @memberof AmqpMetricReporter
    */
-  protected reportMetric(metric: Metric, ctx: MetricSetReportContext<Metric>): Message {
-    return this.options.metricMessageBuilder(ctx.registry, metric, ctx.type, ctx.date, this.buildTags(ctx.registry, metric));
+  protected reportMetric(metric: Metric, ctx: MetricSetReportContext<Metric>): AmqpReportingResult {
+    const tags = this.buildTags(ctx.registry, metric);
+    const message = this.options.metricMessageBuilder(ctx.registry, metric, ctx.type, ctx.date, tags);
+    const routingKey = this.options.routingKeyDeterminator(ctx.registry, metric, ctx.type, ctx.date, tags);
```
Comment (rstiller) [inline lib/metrics/AmqpMetricReporter.ts]: Would it makes sense to make the `routingKeyDeterminator` and `metricMessageBuilder` returning a `Promise<string>` and wait for the completion here? May be adding more future-safe function signature

**Comment by doubret** [comment]
@rstiller i’ll look at your comments this evening, thanks

File: tslint.json (hunk 2 above)
Comment (doubret) [reply tslint.json]: Yes, i agree that 200 is a bit long... but i fail to find a multiline syntax that's pleasant to the eye.
Do not hesitate to modify the code and submit a PR if you want.

File: lib/metrics/index.ts (hunk 1 above)
Comment (doubret) [reply lib/metrics/index.ts]: done (directly on master), thanks

File: lib/metrics/AmqpMetricReporter.ts (hunk 3 above)
Comment (doubret) [reply lib/metrics/AmqpMetricReporter.ts]: done (directly on master), thanks

File: lib/metrics/AmqpMetricReporter.ts (hunk 4 above)
Comment (doubret) [reply lib/metrics/AmqpMetricReporter.ts]: Do you have an example when this could be an asynchronous process ?

Comment (rstiller) [reply lib/metrics/AmqpMetricReporter.ts]: May be someone is collecting some information from a db or cache while building the message or routingKey. Or reading a file and don't want to block the event loop.

**Pull request merged**